source/kernel/src/drivers/emb_display.rs

```diff
@@ -0,0 +1,260 @@
+use crate::{
+    comms::{
+        kchannel::{KChannel, KConsumer},
+        oneshot::Reusable,
+    },
+    registry::{
+        Envelope, KernelHandle, Message, RegisteredDriver, ReplyTo,
+    },
+    Kernel,
+};
+use maitake::sync::Mutex;
+use mnemos_alloc::containers::{HeapArc, HeapArray, HeapFixedVec};
+// use tracing::{debug, warn};
+use uuid::Uuid;
+use embedded_graphics::{
+    pixelcolor::{Gray8, GrayColor},
+    prelude::*,
+    image::ImageRaw,
+};
+
+const BYTES_PER_PIXEL: u32 = 1;
+const DISP_WIDTH: u32 = 319;
+const DISP_HEIGHT: u32 = 239;
+
+// Registered driver
+pub struct EmbDisplay {
+    kernel: &'static Kernel,
+}
+
+// FrameChunk is recieved after client has sent a request for one
+pub struct FrameChunk {
+    frame_id: u16,
+    bytes: HeapArray<u8>,
+    start_x: i32,
+    start_y: i32,
+    width: u32,
+    height: u32,
+}
+
+struct FrameInfo {
+    frame: u16,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum FrameError {
+    DuplicateItem,
+    RegistryFull,
+}
+
+struct DisplayInfo{
+    kernel: &'static Kernel,
+    frames: HeapFixedVec<FrameInfo>,
+    frame_size: usize,
+}
+
+// Client interface to EmbDisplay
+pub struct EmbDisplayHandle {
+    prod: KernelHandle<EmbDisplay>,
+    reply: Reusable<Envelope<Result<Response, FrameError>>>,
+}
+
+pub enum Request {
+    NewFrameChunk { frame_id: u16, start_x: i32, start_y: i32, width: u32, height: u32 },
+}
+
+pub enum Response {
+    FrameChunkAllocated(FrameChunk),
+}
+
+pub enum EmbDisplayError {
+    BufferFull,
+}
+
+struct CommanderTask {
+    cmd: KConsumer<Message<EmbDisplay>>,
+    fmutex: HeapArc<Mutex<DisplayInfo>>, 
+}
+
+// impl EmbDisplay
+impl RegisteredDriver for EmbDisplay {
+    type Request = Request;
+    type Response = Response;
+    type Error = FrameError;
+    const UUID: Uuid = crate::registry::known_uuids::kernel::EMB_DISPLAY;
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum RegistrationError {
+    DisplayAlreadyExists,
+}
+
+// Register the driver instance
+impl EmbDisplay {
+    pub async fn register(
+        kernel: &'static Kernel,
+        max_frames: usize,
+        width: usize,
+        height: usize,
+    ) -> Result<(), ()> {
+        let frames = kernel.heap().allocate_fixed_vec(max_frames).await;
+        let frame_size = width * height * BYTES_PER_PIXEL as usize;
+
+        let imutex = kernel
+            .heap()
+            .allocate_arc(Mutex::new(DisplayInfo { 
+                kernel,
+                frames,
+                frame_size,
+            }))
+            .await;
+        let (cmd_prod, cmd_cons) = KChannel::new_async(kernel, 1).await.split();
+        let commander = CommanderTask {
+            cmd: cmd_cons,
+            fmutex: imutex,
+        };
+
+        kernel.spawn(async move{
+            commander.run().await;
+        }).await;
```

Comment:
tiny style nit, take it or leave it: this creates an unnecessary `async` block. since all we're doing is calling an async fn and awaiting it, this could just be:
```suggestion
        kernel.spawn(commander.run()).await;
```

Reply:
ah ok! what would be the use case for retaining the outer async block? Sorry for the basic question. I'm very new to rust and asynchronous programming! Thanks for the suggestions :)

Reply:
i don't really think there's a reason you would want to do it — since there's no other code in the async block, you can just spawn the method call.